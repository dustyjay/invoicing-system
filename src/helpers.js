export const CONSTANTS = {
  INVOICES: 'invoices',
  SHOW_PREVIEW: 'show-preview',
  USER_DATA: 'user-data',
  INVOICE_NUMBER_LENGTH: 5
};

export const formatInvoiceNumber = number => {
  const num = Math.ceil(number);
  const { length } = `${num}`;
  let zeroDigits = CONSTANTS.INVOICE_NUMBER_LENGTH - length;
  let finalNumber = '';
  while (zeroDigits > 0) {
    finalNumber += '0';
    zeroDigits -= 1;
  }
  return finalNumber + num;
};

export const formatDate = date => {
  if (date) return new Date(date).toDateString();
  return '';
};

export const formatCurrency = value => {
  const formatter = new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: 'USD'
  });

  return formatter.format(value);
};

export const setStorageItem = (key, value) => {
  localStorage.setItem(key, JSON.stringify(value));
};

export const getStorageItem = key => {
  const item = localStorage.getItem(key);
  if (item) return JSON.parse(item);
};

export const removeStorageItem = key => {
  localStorage.removeItem(key);
};
