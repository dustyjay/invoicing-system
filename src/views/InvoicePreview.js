import React from 'react';
import { formatCurrency, formatDate, formatInvoiceNumber } from '../helpers';

const InvoicePreview = ({ invoice, index }) => {
  const getTaxRate = () => {
    const { tax = 0 } = invoice;
    return (tax / 100) * getSumSubTotal();
  };

  const getDiscount = () => {
    const { discount = 0 } = invoice;
    return (discount / 100) * getSumSubTotal();
  };
  const getSumSubTotal = () => {
    const { items } = invoice;
    return items.reduce((acc, { amount, quantity }) => {
      return acc + amount * quantity;
    }, 0);
  };

  const getSumTotal = () => {
    return getSumSubTotal() - getTaxRate() - getDiscount();
  };

  const footerContent = () => {
    const { userData } = invoice;
    const { email, phone, website } = userData;
    const values = [email, phone, website].filter(el => el);
    return values.join(' | ');
  };

  return (
    <section className="preview">
      <div className="preview-box">
        <div className="preview-top">
          <div>
            <div className="preview-data">
              <div className="preview-data__client">
                <p className="preview__name">{invoice.clientName || '--'}</p>
                <p>
                  Date Issued: <strong>{formatDate(invoice.dateIssued)}</strong>
                </p>
                <p>
                  Invoice Number: <strong>#{formatInvoiceNumber(index)}</strong>
                </p>
              </div>
              <div className="preview-data--self">
                <p>{invoice.userData.name || '--'}</p>
                <p>{invoice.userData.address}</p>
                <p>
                  {invoice.userData.town && `${invoice.userData.town}, `}
                  {invoice.userData.city}
                </p>
                <p>{invoice.userData.postalCode}</p>
              </div>
            </div>
            <div>
              <div className="table">
                <div className="table-head">
                  <div className="table-desc">Product Name</div>
                  <div className="table-price">Price</div>
                  <div className="table-qty">Qty</div>
                  <div className="table-subtotal">Subtotal</div>
                </div>
                <hr className="table-head__hr" />
                <div className="table-body">
                  {invoice.items.map(({ name, amount, quantity }, index) => (
                    <div className="table-row" key={index}>
                      <div className="table-desc">{name}</div>
                      <div className="table-price">
                        {formatCurrency(amount)}
                      </div>
                      <div className="table-qty">{quantity}</div>
                      <div className="table-subtotal">
                        {formatCurrency(amount * quantity)}
                      </div>
                    </div>
                  ))}
                </div>
              </div>
            </div>
          </div>
        </div>
        <div className="preview-bottom">
          <div className="preview-meta__box">
            <div className="preview-meta">
              <div>
                <p className="preview-meta__title">Discount:</p>
                <p className="preview-meta__value">{invoice.discount || 0}%</p>
              </div>
              <div className="preview-meta__total">
                {formatCurrency(getDiscount())}
              </div>
            </div>
            <div className="preview-meta">
              <div>
                <p className="preview-meta__title">Tax rate:</p>
                <p className="preview-meta__value">{invoice.tax || 0}%</p>
              </div>
              <div className="preview-meta__total">
                {formatCurrency(getTaxRate())}
              </div>
            </div>
          </div>
          <div>
            <div className="table">
              <div className="table-head">
                <div className="table-bank">Bank Info</div>
                <div className="table-due">Due By</div>
                <div className="table-total">Total Due</div>
              </div>
              <hr className="table-head__hr" />
              <div className="table-body">
                <div className="table-row">
                  <div className="table-bank">
                    <p>
                      Account No:&nbsp;
                      <strong>{invoice.userData.accountNumber || '--'}</strong>
                    </p>
                    <p>
                      Sort Code:&nbsp;
                      <strong>{invoice.userData.sortCode || '--'}</strong>
                    </p>
                  </div>
                  <div className="table-due">
                    {formatDate(invoice.dueDate) || '--'}
                  </div>
                  <div className="table-total">
                    {formatCurrency(getSumTotal())}
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div className="preview-footer">
            <p className="preview-footer__thanks">❤️ Thank you!</p>
            <p>{footerContent()}</p>
          </div>
        </div>
      </div>
    </section>
  );
};

export default InvoicePreview;
