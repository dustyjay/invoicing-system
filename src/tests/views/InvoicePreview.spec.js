import React from 'react';
import { render, cleanup } from '@testing-library/react';
import InvoicePreview from '../../views/InvoicePreview';
import { INVOICES } from '../testData';
import { formatDate } from '../../helpers';

const customRender = (ui, options) =>
  render(ui, {
    wrapper: props => <InvoicePreview {...props} {...options.wrapperProps} />,
    ...options
  });

const INVOICE_INDEX = 0;

describe('Invoice Preview', () => {
  let wrapper;
  let container;
  let getByText;

  beforeEach(() => {
    wrapper = customRender(InvoicePreview, {
      wrapperProps: {
        invoice: INVOICES[INVOICE_INDEX],
        index: INVOICE_INDEX
      }
    });
    container = wrapper.container;
    getByText = wrapper.getByText;
  });

  afterEach(() => {
    cleanup();
  });

  test('Invoice Form matches snapshot', () => {
    expect(container).toMatchSnapshot();
  });

  test("has the right client's data", () => {
    const { clientName = '--', dateIssued, dueDate } = INVOICES[INVOICE_INDEX];
    const clientNameText = getByText(clientName);
    const dateIssuedText = getByText(formatDate(dateIssued));
    const dueDateText = getByText(formatDate(dueDate));
    expect(clientNameText).toBeInTheDocument();
    expect(dateIssuedText).toBeInTheDocument();
    expect(dueDateText).toBeInTheDocument();
  });

  test('has the correct number of items in the invoice', () => {
    const { items } = INVOICES[INVOICE_INDEX];
    const tableRows = container.querySelectorAll('.table-row.items');
    expect(tableRows.length).toEqual(items.length);
  });

  test('has the right user data', () => {
    const { userData = {} } = INVOICES[INVOICE_INDEX];
    const {
      name = '--',
      email,
      phone,
      address,
      town,
      city,
      postalCode,
      accountNumber,
      sortCode,
      website
    } = userData;
    const footerContent = [email, phone, website].filter(el => el).join(' | ');
    const nameText = getByText(new RegExp(name, 'i'));
    const footerText = getByText(new RegExp(footerContent, 'i'));
    const addressText = getByText(new RegExp(address, 'i'));
    const townText = getByText(new RegExp(town));
    const cityText = getByText(new RegExp(city));
    const postalCodeText = getByText(new RegExp(postalCode));
    const accountNumberText = getByText(new RegExp(accountNumber));
    const sortCodeText = getByText(new RegExp(sortCode));
    expect(nameText).toBeInTheDocument();
    expect(footerText).toBeInTheDocument();
    expect(addressText).toBeInTheDocument();
    expect(townText).toBeInTheDocument();
    expect(cityText).toBeInTheDocument();
    expect(postalCodeText).toBeInTheDocument();
    expect(accountNumberText).toBeInTheDocument();
    expect(sortCodeText).toBeInTheDocument();
  });
});
