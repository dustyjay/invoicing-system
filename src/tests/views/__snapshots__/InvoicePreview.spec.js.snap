// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Invoice Preview Invoice Form matches snapshot 1`] = `
<div>
  <section
    class="preview"
  >
    <div
      class="preview-box"
    >
      <div
        class="preview-top"
      >
        <div>
          <div
            class="preview-data"
          >
            <div
              class="preview-data__client"
            >
              <p
                class="preview__name"
              >
                Jurassic Park
              </p>
              <p>
                Date Issued: 
                <strong>
                  Sat May 24 2031
                </strong>
              </p>
              <p>
                Invoice Number: 
                <strong>
                  #
                  00000
                </strong>
              </p>
            </div>
            <div
              class="preview-data--self"
            >
              <p>
                Jane Doe
              </p>
              <p>
                4, Harold road
              </p>
              <p>
                Mowe, 
                Lagos
              </p>
              <p>
                100010
              </p>
            </div>
          </div>
          <div>
            <div
              class="table"
            >
              <div
                class="table-head"
              >
                <div
                  class="table-desc"
                >
                  Product Name
                </div>
                <div
                  class="table-price"
                >
                  Price
                </div>
                <div
                  class="table-qty"
                >
                  Qty
                </div>
                <div
                  class="table-subtotal"
                >
                  Subtotal
                </div>
              </div>
              <hr
                class="table-head__hr"
              />
              <div
                class="table-body"
              >
                <div
                  class="table-row items"
                >
                  <div
                    class="table-desc"
                  >
                    Frontend Development
                  </div>
                  <div
                    class="table-price"
                  >
                    $3,000.00
                  </div>
                  <div
                    class="table-qty"
                  >
                    2
                  </div>
                  <div
                    class="table-subtotal"
                  >
                    $6,000.00
                  </div>
                </div>
                <div
                  class="table-row items"
                >
                  <div
                    class="table-desc"
                  >
                    Backend Development
                  </div>
                  <div
                    class="table-price"
                  >
                    $10,000.00
                  </div>
                  <div
                    class="table-qty"
                  >
                    4
                  </div>
                  <div
                    class="table-subtotal"
                  >
                    $40,000.00
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="preview-bottom"
      >
        <div
          class="preview-meta__box"
        >
          <div
            class="preview-meta"
          >
            <div>
              <p
                class="preview-meta__title"
              >
                Discount:
              </p>
              <p
                class="preview-meta__value"
              >
                0
                %
              </p>
            </div>
            <div
              class="preview-meta__total"
            >
              $0.00
            </div>
          </div>
          <div
            class="preview-meta"
          >
            <div>
              <p
                class="preview-meta__title"
              >
                Tax rate:
              </p>
              <p
                class="preview-meta__value"
              >
                0
                %
              </p>
            </div>
            <div
              class="preview-meta__total"
            >
              $0.00
            </div>
          </div>
        </div>
        <div>
          <div
            class="table"
          >
            <div
              class="table-head"
            >
              <div
                class="table-bank"
              >
                Bank Info
              </div>
              <div
                class="table-due"
              >
                Due By
              </div>
              <div
                class="table-total"
              >
                Total Due
              </div>
            </div>
            <hr
              class="table-head__hr"
            />
            <div
              class="table-body"
            >
              <div
                class="table-row"
              >
                <div
                  class="table-bank"
                >
                  <p>
                    Account No: 
                    <strong>
                      3918789023
                    </strong>
                  </p>
                  <p>
                    Sort Code: 
                    <strong>
                      10482
                    </strong>
                  </p>
                </div>
                <div
                  class="table-due"
                >
                  Sat Jun 14 2031
                </div>
                <div
                  class="table-total"
                >
                  $46,000.00
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="preview-footer"
        >
          <p
            class="preview-footer__thanks"
          >
            ❤️ Thank you!
          </p>
          <p>
            jane@doe.me | +44959270823 | jane-doe.com
          </p>
        </div>
      </div>
    </div>
  </section>
</div>
`;
